0004-predicate-types.md

```diff
@@ -0,0 +1,192 @@
+- Feature Name: predicate_types
+- Start Date: 2022-02-24
+- RFC PR: [dafny-lang/rfcs#0008](https://github.com/dafny-lang/rfcs/pull/0008)
+- Dafny Issue: [dafny-lang/dafny#1680](https://github.com/dafny-lang/dafny/issues/1680)
+
+# Summary
+[summary]: #summary
+
+We introduce the new construct `predicate type`, which acts both like a subset type and a predicate
+```dafny
+predicate type Even(x: int) {
+  x % 2 == 0
+}
+```
+
+which would desugar into:
+```dafny
+predicate Even(x: int) {
+  x % 2 == 0
+}
+
+type Even = x: int | Even(x)
+```
+
+# Motivation
+[motivation]: #motivation
+
+The last snippet above is very common: We define a predicate, and use that predicate to define a subset type.
```

Comment:
Another common idiom is

``` dafny
datatype D_ = ... {
  predicate Valid() { ... }
}
type D = d: D_ | d.Valid()
```

Does `predicate type` offer any advantage here?

Reply:
Very good question. It would, but only if we were able to define types in datatype definitions, e.g.

```
datatype D = ... {
  predicate type Valid() { ... }
}

function f(d: D.Valid) { ... }
```
that would desugar to 

```
datatype D = ... {
  predicate Valid() { ... }
  static type Valid = d: D | d.Valid();
}

function f(d: D.Valid) { ... }
```